File: modules/cpp_workloads/tests/workloads_test.cpp
#include <gtest/gtest.h>

#include <algorithm>
#include <cstddef>
#include <vector>

#include "fragmentation.h"
#include "matmul.h"
#include "pointer_chase.h"
#include "quicksort.h"
#include "retained_growth.h"
#include "string_churn.h"

namespace {

constexpr std::uint64_t SEED = 42;

TEST(Workloads, MatmulLoopOrdersAgree) {
    const std::size_t n = 16;
    const auto a = cpp_workloads::random_matrix(n, SEED);
    const auto b = cpp_workloads::random_matrix(n, SEED + 1);
    // Per-element accumulation order is identical in both variants, so
    // the float results match exactly.
    EXPECT_EQ(cpp_workloads::multiply_ijk(a, b, n), cpp_workloads::multiply_ikj(a, b, n));
}

TEST(Workloads, QuicksortSorts) {
    auto v = cpp_workloads::random_slice(1000, SEED);
    auto expected = v;
    std::ranges::sort(expected);
    cpp_workloads::quicksort(v);
    EXPECT_EQ(expected, v);
}

TEST(Workloads, PointerChaseVisitsEverySlotOnce) {
    const std::size_t n = 97;
    const auto perm = cpp_workloads::build_cycle(n, SEED);
    const std::size_t expected = n * (n - 1) / 2;
    EXPECT_EQ(expected, cpp_workloads::chase_sum(perm));
    EXPECT_EQ(expected, cpp_workloads::array_sum(perm));
}

TEST(Workloads, RetainedGrowthRetainsRequestedBytes) {
    const auto retained = cpp_workloads::grow(8, 1024);
    EXPECT_EQ(8 * 1024, cpp_workloads::retained_bytes(retained));
}

TEST(Workloads, StringChurnBuildsFullString) {
    EXPECT_EQ(20, cpp_workloads::concat(10, "ab").size());
}

TEST(Workloads, FragmentationKeepsEveryOtherBlockDoubled) {
    const auto [survivors, stats] = cpp_workloads::fragment(9, SEED);
    EXPECT_EQ(5, stats.survivors);
    EXPECT_EQ(survivors.size(), stats.survivors);
    std::size_t total = 0;
    for (const auto& block : survivors) {
        // Doubled from the initial 512..8191 draw.
        EXPECT_GE(block.size(), 2 * 512);
        EXPECT_LT(block.size(), 2 * 8192);
        total += block.size();
    }
    EXPECT_EQ(total, stats.live_bytes);
}

}  // namespace

File: modules/java_workloads/tests/WorkloadsTest.java
package monorepo.workloads;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.Arrays;

import org.junit.jupiter.api.Test;

class WorkloadsTest {
    private static final long SEED = 42;

    @Test
    void matmulLoopOrdersAgree() {
        int n = 16;
        double[] a = Matmul.randomMatrix(n, SEED);
        double[] b = Matmul.randomMatrix(n, SEED + 1);
        // Per-element accumulation order is identical in both variants, so
        // the float results match exactly.
        assertArrayEquals(Matmul.multiplyIjk(a, b, n), Matmul.multiplyIkj(a, b, n));
    }

    @Test
    void quicksortSorts() {
        long[] v = Quicksort.randomSlice(1000, SEED);
        long[] expected = Arrays.copyOf(v, v.length);
        Arrays.sort(expected);
        Quicksort.quicksort(v);
        assertArrayEquals(expected, v);
    }

    @Test
    void pointerChaseVisitsEverySlotOnce() {
        int n = 97;
        int[] perm = PointerChase.buildCycle(n, SEED);
        long expected = (long) n * (n - 1) / 2;
        assertEquals(expected, PointerChase.chaseSum(perm));
        assertEquals(expected, PointerChase.arraySum(perm));
    }

    @Test
    void retainedGrowthRetainsRequestedBytes() {
        byte[][] retained = RetainedGrowth.grow(8, 1024);
        assertEquals(8 * 1024, RetainedGrowth.retainedBytes(retained));
    }

    @Test
    void stringChurnBuildsFullString() {
        assertEquals(20, StringChurn.concat(10, "ab").length());
    }
}
